python/tritonserver/_api/_dlpack.py

```diff
@@ -0,0 +1,272 @@
+
+
+def triton_to_dlpack_dtype(dtype):
+    if dtype == "BOOL":
+        type_code = DLDataTypeCode.kDLBool
+        bits = 1
+    elif dtype == "INT8":
+        type_code = DLDataTypeCode.kDLInt
+        bits = 8
+    elif dtype == "INT16":
+        type_code = DLDataTypeCode.kDLInt
+        bits = 16
+    elif dtype == "INT32":
+        type_code = DLDataTypeCode.kDLInt
+        bits = 32
+    elif dtype == "INT64":
+        type_code = DLDataTypeCode.kDLInt
+        bits = 64
+    elif dtype == "UINT8":
+        type_code = DLDataTypeCode.kDLUInt
+        bits = 8
+    elif dtype == "UINT16":
+        type_code = DLDataTypeCode.kDLUInt
+        bits = 16
+    elif dtype == "UINT32":
+        type_code = DLDataTypeCode.kDLUInt
+        bits = 32
+    elif dtype == "UINT64":
+        type_code = DLDataTypeCode.kDLUInt
+        bits = 64
+    elif dtype == "FP16":
+        type_code = DLDataTypeCode.kDLFloat
+        bits = 16
+    elif dtype == "FP32":
+        type_code = DLDataTypeCode.kDLFloat
+        bits = 32
+    elif dtype == "FP64":
+        type_code = DLDataTypeCode.kDLFloat
+        bits = 64
+    elif dtype == "BF16":
+        type_code = DLDataTypeCode.kDLBfloat
+        bits = 16
+    elif dtype == "BYTES":
+        _raise_error("DLPack currently doesn't suppose BYTES type")
+    else:
+        _raise_error(
+            "Can not covert unknown data type '{}' to DLPack data type".format(dtype)
+        )
+    return DLDataType(type_code, bits, 1)
+
+
+def is_contiguous_data(
+    ndim: ctypes.c_int,
+    shape: ctypes.POINTER(ctypes.c_int64),
+    stride: ctypes.POINTER(ctypes.c_int64),
+):
+    # If 'stride' doesn't capture valid value
+    if (stride is None) or (not bool(stride)):
+        return True
+    calculated_stride = 1
+    # iterate stride in reverse order [ndim-1, -1)
+    for i in reversed(range(ndim)):
+        if stride[i] != calculated_stride:
+            return False
+        calculated_stride *= shape[i]
+    return True
+
+
+def get_byte_size(
+    dtype: DLDataType, ndim: ctypes.c_int, shape: ctypes.POINTER(ctypes.c_int64)
+):
+    element_byte_size = dtype.bits * dtype.lanes // 8  # Assume 8 bits in a byte
+    for i in range(ndim):
+        element_byte_size *= shape[i]
+    return element_byte_size
+
+
+def get_dlpack_capsule(dlpack_obj, stream=None):
+    # Extract PyCapsule of the DLPack object
+    if hasattr(dlpack_obj, "__dlpack__"):
+        if not hasattr(dlpack_obj, "__dlpack_device__"):
+            _raise_error(
+                "DLPack expects '__dlpack_device__' if '__dlpack__' has been defined"
+            )
+        device = dlpack_obj.__dlpack_device__()
+        # Have to condition on the device type as, using numpy as example,
+        # some DLPack implementation doesn't accept 'stream' as arguments
+        if device != DLDeviceType.kDLCUDA:
+            return dlpack_obj.__dlpack__()
+        else:
+            return dlpack_obj.__dlpack__(stream)
```

Comment:
I think here we may need to synchronize the default stream here. You've probably discussed this with @oandreeva-nv offline. This is an important issue that needs to be fixed before merging the PR.

Reply:
after conversation with @oandreeva-nv synchronize on legacy default stream (1)

Reply:
Sorry was late to the game, yes, we've discussed this offline